Basic ClientHeadersFactory documentation (#13891)
quarkusio/quarkus: docs/src/main/asciidoc/rest-client.adoc

More details about this can be found in https://smallrye.io/smallrye-mutiny/#_uni_and_multi[the Mutiny documentation].

== Custom header support
The mp-rest client allows registering factory to amend requests headers using `@RegisterClientHeaders` annotation.

Let's see it in action by adding a `RequestUUIDHeaderFactory` method in our `CountriesService` REST interface. The code should look like:

[source, java]
----
package org.acme.rest.client;

import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
import org.jboss.resteasy.annotations.jaxrs.PathParam;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import java.util.concurrent.CompletionStage;
import java.util.Set;

@Path("/v2")
@RegisterRestClient
@RegisterClientHeaders(RequestUUIDHeaderFactory.class)
public interface CountriesService {

    @GET
    @Path("/name/{name}")
    @Produces("application/json")
    Set<Country> getByName(@PathParam String name);

    @GET
    @Path("/name/{name}")
    @Produces("application/json")
    CompletionStage<Set<Country>> getByNameAsync(@PathParam String name);
}
----

> And the RequestUUIDHeaderFactory` :

Reviewer: ```suggestion
And the `RequestUUIDHeaderFactory` would look like:
```

[source, java]
----
package org.acme.rest.client;

import org.eclipse.microprofile.rest.client.ext.ClientHeadersFactory;
import org.jboss.resteasy.specimpl.MultivaluedMapImpl;

import javax.enterprise.context.ApplicationScoped;
import javax.ws.rs.core.MultivaluedMap;
import java.util.UUID;

@ApplicationScoped
public class RequestUUIDHeaderFactory implements ClientHeadersFactory {

    @Override

>     public MultivaluedMap<String, String> update(MultivaluedMap<String, String> multivaluedMap, MultivaluedMap<String, String> multivaluedMap1) {

Reviewer: ```suggestion
    public MultivaluedMap<String, String> update(MultivaluedMap<String, String> incomingHeaders, MultivaluedMap<String, String> clientOutgoingHeaders) {
```

        MultivaluedMap<String, String> result = new MultivaluedMapImpl<>();
        result.add("X-request-uuid", UUID.randomUUID().toString());
        return result;
    }
}
----

> Any CDI bean matching the declared factory will be picked-up and can of course benefit from quarkus features.

Reviewer: ```suggestion
Any CDI bean matching the declared factory class will be picked up and can of course benefit from the usual CDI infrastructure.
```
